src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -46,9 +54,16 @@ public TElement[] ToArray()
 
             TElement[] array = new TElement[count];
             int[] map = SortedMap(buffer);
```

Comment:
That'd return the map to the pool, even when it shouldn't. If you wanted to use finally (I don't know, for example, what the cost difference of `catch` and `finally` is) it would have to be:

```C#
int[] map = ComputeMap(elements, count);
bool success = false;
try
{
    QuickSort(map, 0, count - 1);
    success = true;
}
finally
{
    if (!success)
    {
        ArrayPool<int>.Shared.Return(map);
     }
}

return map;
```

Maybe C# 8 or C# 9 will let us just do:

```C#
int[] map = ComputeMap(elements, count);
bool success = false;
try
{
    QuickSort(map, 0, count - 1);
    success = true;
}
fault
{
    ArrayPool<int>.Shared.Return(map);
}
return map;
```
:wink: